pkg/sql/alter_column_type.go

```diff
@@ -231,14 +222,75 @@ func alterColumnTypeGeneral(
 
 	shadowColName := sqlbase.GenerateUniqueConstraintName(col.Name, nameExists)
 
-	// The default computed expression is casting the column to the new type.
-	newComputedExpr := tree.CastExpr{
-		Expr:       &tree.ColumnItem{ColumnName: tree.Name(col.Name)},
-		Type:       toType,
-		SyntaxMode: tree.CastShort,
+	var newColComputeExpr *string
+	// oldCol still needs to have values written to it in case nodes read it from
+	// it with a TableDescriptor version from before the swap.
+	// To achieve this, we make oldCol a computed column of newCol using
+	// inverseExpr.
+	// If a USING EXPRESSION was provided, we cannot generally invert
+	// the EXPRESSION to insert into the old column and thus will force an
+	// error using the computed expression. Any inserts into the new column
+	// will fail until the old column is dropped.
+	var inverseExpr string
+	if Using != nil {
+		// Ensure the USING EXPRESSION returns the new type.
+		// Replace column references with typed dummies to allow typechecking.
+		replacedExpr, _, err := replaceVars(tableDesc, Using)
+		if err != nil {
+			return err
+		}
+
+		_, err = sqlbase.SanitizeVarFreeExpr(
+			ctx, replacedExpr, toType, "ALTER COLUMN TYPE USING EXPRESSION", &params.p.semaCtx, true, /* allowImpure */
+		)
+		if err != nil {
+			return err
+		}
+		s := tree.Serialize(Using)
+		newColComputeExpr = &s
+
+		insertedValToString := tree.CastExpr{
+			Expr:       &tree.ColumnItem{ColumnName: tree.Name(col.Name)},
+			Type:       types.String,
+			SyntaxMode: tree.CastShort,
+		}
+		insertedVal := tree.Serialize(&insertedValToString)
+		// Set the computed expression to use crdb_internal.force_error() to
+		// prevent writes into the column. Whenever the new column is written to
+		// crdb_internal.force_error() will cause the write to error out.
+		// This prevents the column going ALTER COLUMN TYPE to be unable to write
+		// to until the old column is dropped.
+		// This is safe to do so because after the column swap, the old column
+		// should become "read-only".
+		// The computed expression uses the old column to trigger the error
+		// by using the string 'Tried to insert <value> into <column name>'.
+		errMsg := fmt.Sprintf(
+			"'Column %s is undergoing the ALTER COLUMN TYPE USING EXPRESSION "+
+				"schema change. Inserts are not supported until the schema change is "+
+				"finalized'", col.Name)
+		failedInsertMsg := fmt.Sprintf(
+			"'\nTried to insert ', %s, ' into %s'", insertedVal, col.Name,
+		)
+		inverseExpr = fmt.Sprintf(
+			"crdb_internal.force_error('03000', concat(%s, %s))",
+			errMsg, failedInsertMsg)
+	} else {
+		// The default computed expression is casting the column to the new type.
+		newComputedExpr := tree.CastExpr{
+			Expr:       &tree.ColumnItem{ColumnName: tree.Name(col.Name)},
```

Comment:
Does anything weird happen here by setting the name of the column in the expression here -- I know we pick new names for columns in the mutation.

Reply:
I think this is the correct behaviour since both the computed expressions should be based on the name of the current column. When backfilling the old column, we use the current column and then the "inverse" expression is used after the swap so the new column will now be named as the original column name.